#166575111 Implement Favorite and Unfavorite Book

henryade opened the pull request:
#### What does this PR do?
adds the favorite and unfavorite book functionality
#### Description of Task to be completed?
- create a favourite book model
- create author, book and favorite book seeders
- added new user with the role
- create route and user controller
 - test endpoint with all use cases
#### How should this be manually tested?
- clone the project 
- install all dependencies
- run test
OR
- start server and test manually by
- run npm run db:rollback
- sign in with a patron account and copy the token
- open postman and go the `http://localhost:{PORT}/api/v1/book/:bookId` and send a post request to favorite book. paste the token as a query or in the headers
- use 1 as bookId for test purposes
- send a delete request with the same url to unfavorite book
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166575111](https://www.pivotaltracker.com/story/show/166575111)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

KvNGCzA commented on server/controllers/UserController.js:
```diff
@@ -0,0 +1,74 @@
+    try {
+      const [favoritebook, book, param] = await UserController.findBook(request);
+      if (!book) return responseMessage(response, 404, { message: 'book not found' });
+      if (favoritebook) return responseMessage(response, 400, { message: 'book already in your favourites' });
```
since the book already exists as a favourite, I think this should be a `conflict` status code, what do you think?

henryade replied:
done

allebd commented on server/database/migrations/20190610121629-create-favorite-book.js:
```diff
@@ -0,0 +1,42 @@
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('FavoriteBooks');
+  }
+};
```
New line required

henryade replied:
done

allebd commented:
No swagger documentation and your PR does not state the URL to favourite or unfavourite a book in the description above

allebd commented on server/controllers/UserController.js:
```diff
@@ -0,0 +1,74 @@
+      if (favorite) {
+        return responseMessage(response, 200, {
+          status: 'success',
+          message: 'favorited successfully'
```
can I get **data** information of the book successfully favorited

henryade replied:
done

Adekoreday reviewed (changes requested):
swagger documentation required

done

allebd commented on test/server/routes/user.test.js:
```diff
@@ -0,0 +1,148 @@
+        expect(response.body.message).to.equal('unfavorited successfully');
+        expect(response.body.book).to.be.a('object');
+        expect(response.body.book).to.have.property('authorName');
+        expect(response.body.book).to.have.property('authorName');
```
Duplicate expect(response.body.book).to.have.property('authorName')'

henryade replied:
changed

henryade commented:
> No swagger documentation and your PR does not state the URL to favourite or unfavourite a book in the description above

Done

allebd reviewed (commented):
Have tested it, nice work but no swagger documentation

henryade commented:
> Have tested it, nice work but no swagger documentation

check very well, i added it

Seen

Lundii reviewed (commented):
How to test the route on your PR body is not well detailed, I really don't know what to test and how to test it

-Done

Ukhu commented on server/controllers/UserController.js:
```diff
@@ -0,0 +1,94 @@
+      const favorite = await FavoriteBook.create(param);
+
+      if (favorite) {
+        return responseMessage(response, 200, {
```
201 statusCode is recommended here since a new instance of FavouriteBook model was created.

henryade replied:
I believed 200 was okay since it's being added/removed to/from favorite list. Will change it

henryade replied:
done

Ukhu commented on server/database/seeders/20190530205345-demo-users.js:
```diff
@@ -92,6 +92,17 @@ module.exports = {
       createdAt: new Date(),
       updatedAt: new Date()
+    }, {
+      id: 5445,
```
I think we agreed 'id' should be removed since it's auto-incremented

henryade replied:
Okay

henryade replied:
done

Ukhu commented on server/database/seeders/20190531172358-demo-userroles.js:
```diff
@@ -24,6 +24,12 @@ module.exports = {
     createdAt: new Date(),
     updatedAt: new Date()
+  }, {
+    id: 7866,
```
same here too. 'id' should be removed

henryade replied:
Okay

henryade replied:
done

Ukhu commented on server/middlewares/index.js:
```diff
@@ -4,6 +4,7 @@ import checkForErrors from './checkForErrors';
 import AuthorValidator from './AuthorValidator';
 import authorizeUser from './authorizeUser';
 import BookValidator from './BookValidator';
+import FavoriteBookValidator from './FavoriteBookValidator';
```
instead of creating another file for FavouriteBookValidator, won't it be better for the validator to be a method in the BookValidator class since it's still related to books?

henryade replied:
Okay, I will change this.

henryade replied:
done

Ukhu commented on test/server/routes/__mocks__/userData.js:
```diff
@@ -120,3 +120,28 @@ export const wrongLengthInput = {
+  password: 'g4g5j6lkl4k9',
+};
+
+export const newBook = {
```
Isn't it better for the data relating to books to be in a separate file, say 'bookData' ?

henryade replied:
I put it here because it relates to the user controller and user route.

Nennyfills reviewed (commented):
This branch Heroku deployment is failing

The heroku development branch is falling because the environment variables have not been added and probably some other problems

ad3ola committed:
implement favourite and unfavourite Book
create favourite book model
create author, book and favorite book seeders
added new user with role
create route and user controller
test end point with all usecase
[Delivers #166575111]

allebd commented on server/controllers/BookController.js:
```diff
@@ -17,25 +17,26 @@ class BookController {
+      const message = `book successfully added${newAuthor[1] ? ' and author created' : ''}`;
       return existingBook[1]
-        ? responseMessage(response, 201, { status: 'success', message, book: { ...dataValues, author: authorName } })
+        ? responseMessage(response, 201, { status: 'success', message, book: { ...dataValues, author: newAuthor[0].id } })
```
Your response here should be the author's name, not id

henryade replied:
Speak to biola. And I thought we were asked to change it to id by mengere

allebd replied:
Okay, what he meant was for data entry, imagine a drop down, id would be the value, not in the response

allebd commented on server/controllers/BookController.js:
```diff
@@ -17,25 +17,26 @@ class BookController {
+      body: {
+        title, description, isbn, price, yearPublished, authorName
+      }
+    } = request;
```
What of if you have several authors to a book, how would that be tackled

henryade replied:
I am not the one that implemented this feature. Books asked me to help him change somethings in my PR. Please speak to books.

allebd replied:
kindly revert back so biola can create a bug fix for this or you can assist him

allebd commented on server/database/migrations/20190529204542-create-books.js:
```diff
@@ -19,7 +19,7 @@ module.exports = {
       isbn: {
         allowNull: false,
-        type: Sequelize.TEXT
+        type: Sequelize.BIGINT
```
ISBN is not BIGINT, it consists of four parts for 10 digits or 5 parts for 13 digits, separated by a dash (-)

henryade replied:
Biola.

allebd replied:
kindly revert back so biola can create a bug fix for this or you can assist him

allebd commented on server/database/models/book.js:
```diff
@@ -2,10 +2,9 @@ export default (sequelize, DataTypes) => {
     title: DataTypes.STRING,
     description: DataTypes.TEXT,
-    isbn: DataTypes.TEXT,
+    isbn: DataTypes.BIGINT,
```
ISBN not BIGINT

henryade replied:
Biola also

allebd replied:
kindly revert back so biola can create a bug fix for this or you can assist him

allebd commented on server/middlewares/BookValidator.js:
```diff
@@ -54,7 +54,9 @@ export default class BookValidator {
       .isISBN()
-      .withMessage('isbn is not valid');
+      .withMessage('isbn is not valid')
+      .isInt({ min: 1000000000, max: 9999999999999 })
```
This is not an ISBN, for more information check here https://en.wikipedia.org/wiki/International_Standard_Book_Number

henryade replied:
Speak to biola.

allebd replied:
kindly revert back so biola can create a bug fix for this or you can assist him

allebd commented on server/routes/user.js:
```diff
@@ -2,13 +2,25 @@ import express from 'express';
+// Favorite Book
+user.post('/favorites/book/:bookId', verifyToken, authorizeUser(['patron']), FavoriteBookValidation(), favoriteBook);
+// Unfavorite Book
+user.delete('/favorites/book/:bookId', verifyToken, authorizeUser(['patron']), FavoriteBookValidation(), unfavoriteBook);
```
Delete should be a query, not a param

henryade replied:
Query how?

allebd replied:
You are correct, it can be used both has a param or query string
